src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                    data: r,
+                    styles: comboBoxOptionStyles
+                } as IComboBoxOption);
+            }
+        });
+        return newOptions;
+    };
+
+    // reset the dropdown options after fetching data, first options are resources, last items are the freeform added ones
+    useEffect(() => {
+        let newOptions: Array<IComboBoxOption> = [];
+        if (resourcesState.adapterResult.getCatastrophicError()) {
+            resourcesRef.current = null;
+            setError(resourcesState.adapterResult.getCatastrophicError());
+            newOptions = additionalOptions || [];
+            if (
+                selectedOption &&
+                !additionalOptions?.find((o) => o.text === selectedOption.text)
+            ) {
+                newOptions = newOptions.concat([selectedOption]);
+            }
+        } else if (!resourcesState.adapterResult.hasNoData()) {
+            const resources: Array<IAzureResource> =
+                resourcesState.adapterResult.result?.data;
+            resourcesRef.current = resources;
+
+            if (onResourcesLoaded) {
+                onResourcesLoaded(resources);
+            }
+
+            newOptions = mergeOptionsWithProps(
+                optionsFromResources(),
+                additionalOptions,
+                selectedOption
+            );
+        } else {
+            resourcesRef.current = [];
+            // by default on mount, set the options as additional options and/or selected option for the dropdown
+            if (additionalOptions) {
+                newOptions = additionalOptions;
+            }
+            if (
+                selectedOption &&
+                !additionalOptions?.find((o) => o.text === selectedOption.text)
+            ) {
+                newOptions = newOptions.concat([selectedOption]);
+            }
+        }
+
+        setOptions(newOptions);
+    }, [resourcesState.adapterResult]);
+
+    useEffect(() => {
+        if (selectedOptionProp) {
+            setSelectedOption({
+                key: selectedOptionProp,
+                text: selectedOptionProp,
+                styles: comboBoxOptionStyles
+            });
+        } else {
+            setSelectedOption(null);
+        }
+    }, [selectedOptionProp]);
+
+    useEffect(() => {
+        if (selectedOption) {
+            const selectedKey = options?.find(
+                (option) =>
+                    option.text === parsedOptionText(selectedOption.text)
+            )?.key;
+            setSelectedKey(selectedKey);
+        } else {
+            setSelectedKey(null);
+        }
+    }, [selectedOption]);
+
+    useEffect(() => {
+        setAdditionalOptions(
+            additionalOptionsProp?.map((aO) => ({
+                key: aO,
+                text: aO,
+                styles: comboBoxOptionStyles
+            })) || []
+        );
+    }, [additionalOptionsProp]);
+
+    useEffect(() => {
+        setOptions(
```

Comment:
this nesting of functions feels wrong. They each have many internal side effects and logics. Not sure it can be helped at this point with a lot of churn, but wanted to call it out.